### ISSUE TITLE
Consider making advanced service fields non-optional

### ISSUE THREAD

**Issue by Oaphi** [mtgto/dts-google-apps-script-advanced#13]
Stemming from [this Stack Overflow question](https://stackoverflow.com/q/68747314/11407695), this is a request to consider switching advanced service fields from optional to non-optional. I have a PR ready that will make the change if need be, but wanted to start a preliminary discussion as maybe there is a reason why they are generated this way.

Currently, the `Field` class prints *all* fields as optional unconditionally when the `toString` method is invoked:

https://github.com/mtgto/dts-google-apps-script-advanced/blob/ffd96078229a1c63e75b5ab3ae34fef83b4296d5/src/parser.ts#L135

This is:

- inconsistent with the [Node.js client](https://github.com/googleapis/google-api-nodejs-client/tree/main/src/apis) definitions
- leading to deep optional chaining `Sheets.Spreadsheets?.DeveloperMetadata?.get`

If there is no blocking reason why those the fields should not be non-optional, it is proposed to replace the `?:` with `:` when printing fields or consider adding an `optional` parameter to the `Field` constructor with a `false` default parameter (likely safer).

**Comment by mtgto** [comment]
@Oaphi 

> Currently, the Field class prints all fields as optional unconditionally when the toString method is invoked

Yes.

> If there is no blocking reason why those the fields should not be non-optional, it is proposed to replace the `?:` with `:` when printing fields or consider adding an optional parameter to the Field constructor with a false default parameter (likely safer).

It is not easy.

My program generates type definition from json data which Google creates for web browser editor.
This json describe name of fields, properties and methods.
Unfortunetely, the json does not contains whether field is optional or not.

If you automatically replace `?:` to `:` in `toString`, It might generate unwanted field type definition.

> adding an optional parameter to the Field constructor with a false default parameter

I can adopt it, but I can't find which Field is not optional from `json` now.

IMO... the better way is to refer to "Google's Node.js client definition" while generate type definition of advanced services of GAS.
(It is hard, of couse.)

----

Another suggesntion is adding exceptional rule to toString when `name` is Uppercased.
Like `AdminDirectory.Domains`, name consists of `Camecase.Camelcase` style, this field is mostly non-optional.

**Comment by Oaphi** [comment]
Heya, @mtgto!

Thank you for responding on this so soon - I suspected this is erroring on the side of caution because of the information made available in Google's JSON. 

An additional issue is that we are likely going to lose the ability to get that JSON in the first place once Google drops support for the old editor entirely - from what I gathered, the new dialog does not load the definitions as described in [this issue](https://github.com/mtgto/dts-google-apps-script-advanced/issues/5#issuecomment-546321498).

Maybe it is time to switch to the [Discovery API](https://developers.google.com/discovery/v1/reference/apis/list?apix_params=%7B%22name%22%3A%22admin%22%7D) to get the definitions? The discovery document available at https://admin.googleapis.com/$discovery/rest?version=directory_v1 follows the JSON schema, so should not be hard to parse when generating the definitions. I understand that this would lead to an extensive rework, so can offer any help needed and make a PR (won't be very soon, but we are in no rush, methinks).

As a temporary fix,  making camel-cased fields required does indeed look like a safe option, at first glance it does not seem like it should cause issues, and the regenerated types will be reviewed either way.

Please, let me know what your thoughts are. I can start looking into the abovementioned options in the meantime
